python/test/test_api.py

```diff
@@ -71,6 +71,119 @@ def test_create_request(self):
 
 
 class AllocatorTests(unittest.TestCase):
+    class MockMemoryAllocator(tritonserver.MemoryAllocator):
+        def __init__(self):
+            pass
+
+        def allocate(self, *args, **kwargs):
+            raise Exception("foo")
+
+    @pytest.mark.skipif(cupy is None, reason="Skipping gpu memory, cupy not installed")
+    def test_memory_fallback_to_cpu(self):
+        server = tritonserver.Server(server_options).start(wait_until_ready=True)
+
+        self.assertTrue(server.ready())
+
+        allocator = tritonserver.default_memory_allocators[tritonserver.MemoryType.GPU]
+
+        del tritonserver.default_memory_allocators[tritonserver.MemoryType.GPU]
+
+        server.load(
+            "test",
+            {
+                "config": json.dumps(
+                    {
+                        "backend": "python",
+                        "parameters": {
+                            "decoupled": {"string_value": "False"},
+                            "request_gpu_memory": {"string_value": "True"},
```

Comment:
If you set the `requested_output_memory_type = GPU`, would python backend also give you back a GPU buffer without having to use this custom parameter?

Reply:
My understanding is that the backend requests memory from the client for each output tensor - and it can query the allocator to know what type of memory the client will provide. So the requested memory type is the backend requesting a memory type from the client - and exactly the behavior we are relying on for testing. The python backend (turns out) will request GPU memory from the client if the output tensor produced by the model is on the GPU .... another way to think of it is that the client / response allocator always has the final say in the memory type - the backend can only 'request' it.